_posts/2016-05-30-pitest-with-java.md

```diff
@@ -0,0 +1,263 @@
+	List<Passenger> passengers = new ArrayList<>();
+	Passenger adultPassenger1 = new Passenger(20);
+	Passenger childPassenger2 = new Passenger(12);
+	passengers.add(adultPassenger1);
+	passengers.add(childPassenger2);
+	double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+	assertEquals((ADULT_TICKET_PRICE + CHILD_TICKER_PRICE), price, 0);
+}
+{% endhighlight %}
+
+Ok, we covered our edge cases, however if you'll execute pitest plugin once again you'll see that we still have two mutations which survived.
+
+<img src="/images/blog/posts/pitest-with-java/survivedMutations.jpg" />
+
+Take a look into this code:
+
+{% highlight java %}
+if (childrenCounter > 1 && adultCounter > 1) {
+    result = (1 - FAMILY_DISCOUNT) * totalPrice;
+} else {
+    result = totalPrice;
+}
+{% endhighlight %}
+
+Our edge case test scenario passes one adult and one child to this method. Due to AND operator we can change our code to:
+{% highlight java %}
+if (childrenCounter >= 1 && adultCounter > 1) {
+    result = (1 - FAMILY_DISCOUNT) * totalPrice;
+} else {
+    result = totalPrice;
+}
+{% endhighlight %}
+
+and all test all still green, because second condition *adultCounter > 1* change the result of this statement to false. The same behaviour we can observe when we'll change *adultCounter > 1* to *adultCounter >= 1*.
+To cover this cases we should replace our **shouldNotCalculatePriceForFamily** with two new tests:
+{% highlight java %}
+@Test
+public void shouldNotCalculatePriceForFamilyEdgeCaseWithAdults() {
+	List<Passenger> passengers = new ArrayList<>();
+	Passenger adultPassenger1 = new Passenger(20);
+	Passenger adultPassenger2 = new Passenger(20);
+	Passenger childPassenger1 = new Passenger(12);
+	passengers.add(adultPassenger1);
+	passengers.add(adultPassenger2);
+	passengers.add(childPassenger1);
+	double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+	assertEquals((2 * ADULT_TICKET_PRICE + CHILD_TICKER_PRICE), price, 0);
+}
+
+@Test
+public void shouldNotCalculatePriceForFamilyEdgeCaseWithChildren() {
+	List<Passenger> passengers = new ArrayList<>();
+	Passenger adultPassenger1 = new Passenger(20);
+	Passenger childPassenger1 = new Passenger(12);
+	Passenger childPassenger2 = new Passenger(12);
+	passengers.add(adultPassenger1);
+	passengers.add(childPassenger1);
+	passengers.add(childPassenger2);
+	double price = calculator.calculatePrice(passengers, ADULT_TICKET_PRICE, CHILD_TICKER_PRICE);
+	assertEquals((ADULT_TICKET_PRICE + 2 * CHILD_TICKER_PRICE), price, 0);
+}
+{% endhighlight %}
+
+After this we finally reached 100% mutation coverage.
+
+# Summary
+Now you know how to use mutation testing to see if your tests are good. You also know that line coverage doesn't mean that we have covered all cases. Hope you enjoyed and you'll start to use mutation testing. All code which was used is available on [repository](https://github.com/BeyondScheme/pitest-java-example).
```

Comment:
To sum up, after this reading you should learn how to use mutation testing to see if your tests are good.